docs_src/design/adr/security/0026-authentication.md

```diff
@@ -0,0 +1,296 @@
+
+The AS-IS Architecture figure below depicts the current state of
+microservice communication security as of the EdgeX Levski release,
+when security is enabled:
+
+![AS-IS Architecture](0026-as-is.jpg)
+
+As shown in the diagram,
+many of the foundational services used by EdgeX Foundry
+have already been secured:
+
+* Communication with EdgeX's secret store, as implemented by
+  Hashicorp Vault, is secured over a local HTTP socket with
+  token-based authentication.  An access control list limits access
+  to the keyspace of the key value store.
+
+* Communication with EdgeX's service registry and configuration provider,
+  as implemented by Hashicorp Consul, is secured over a local HTTP
+  socket with token-based authentication,
+  with the token being mediated by Hashicorp Vault.
+  An access control list limits access to the keyspace of
+  the configuration store.
+
+* Communication with EdgeX's default database, Redis, is secured using
+  username/password authentication, with the password stored
+  in Hashicorp Vault.  An access control list limits the commands
+  that clients are allowed to issue to the server.
+
+External access to EdgeX microservices has also been secured.
+EdgeX microservices only bind to local ports,
+and are only exposed externally through a Kong API gateway.
+This gateway is configured to use TLS 1.3,
+using JWT authentication using public key cryptography.
+All external requests are filtered at the API gateway.
+URL rewriting is used to concentrate microservices
+on a single HTTP-accessible port.
+
+Behind the proxy, it is not possible to verify Kong
+as the origin of local network traffic because mutual-auth TLS
+is not supported in the open source version of Kong.
+Although the Kong JWT plugin will set request headers
+on the backend request that identify the caller,
+there is no mechanism by which Kong can prove to a
+backend service that it was the component that
+performed the authentication step.
+Even though the original JWT passes through the proxy,
+the Kong authentication plugins do not expose
+token introspection endpoints that the backend service
+could use to check token validity independently.
+
+The consequence of having an API gateway that performs
+all microservice authentication is that communication
+between EdgeX microservices running behind the API gateway
+are not authenticated in any way.
+EdgeX microservices are unable to distinguish
+malicous traffic that has evaded the API gateway
+from legitimate microservice traffic.
+
+
+## Decision
+
+This ADR expands the use of Vault to include
+issuing and validating [JWTs](https://www.rfc-editor.org/rfc/rfc7519)
+for use in microservice authentication among EdgeX services.
+Vault, and not Kong, owns JWTs issuance and validation
+in the new propsal.
+
+Having been relieved of JWT management responsibility,
+it is now possible to replace the Kong API gateway and its associated
+Postgres database with a much simpler (and smaller) NGINX reverse proxy.
```

Comment:
Me likey!!  Is Postgres only used for JWT persistence and retrieval?

Reply:
Postgres is the full configuration store for Kong.  Postgres doesn't store the JWT's themselves, but the validation key for JWT's.  Additionally, it stores the URL rewriting rules, any custom headers we want to inject when forwarding to backend services, and any custom TLS certificates we want to serve on the public endpoint.  However, all of these can easily be put in a configuration file.